arangod/Aql/IResearchViewOptimizerRules.cpp

```diff
@@ -1111,6 +1111,94 @@ void handleConstrainedSortInView(Optimizer* opt,
   }
 }
 
+void immutableSearchCondition(Optimizer* opt,
+                              std::unique_ptr<ExecutionPlan> plan,
+                              OptimizerRule const& rule) {
+  TRI_ASSERT(plan && plan->getAst());
+
+  // ensure 'Optimizer::addPlan' will be called
+  bool modified = false;
+  irs::Finally addPlan = [opt, &plan, &rule, &modified]() noexcept {
+    opt->addPlan(std::move(plan), rule, modified);
+  };
+
+  if (!plan->contains(ExecutionNode::ENUMERATE_IRESEARCH_VIEW)) {
+    return;
+  }
+
+  containers::SmallVector<ExecutionNode*, 8> viewNodes;
+  plan->findNodesOfType(viewNodes, ExecutionNode::ENUMERATE_IRESEARCH_VIEW,
+                        true);
+  VarSet vars;
+  VarSet mutableVars;
+  for (auto* node : viewNodes) {
+    TRI_ASSERT(node);
+    TRI_ASSERT(ExecutionNode::ENUMERATE_IRESEARCH_VIEW == node->getType());
+    auto& view = *ExecutionNode::castTo<IResearchViewNode*>(node);
+    auto const* condition = &view.filterCondition();
+    if (isFilterConditionEmpty(condition) || !view.scorers().empty() ||
+        !isInInnerLoopOrSubquery(view)) {
+      continue;
+    }
+    vars.clear();
+    Ast::getReferencedVariables(condition, vars);
+    vars.erase(&view.outVariable());
+    for (auto* var : vars) {
+      auto* setter = plan->getVarSetBy(var->id);
+      if (!setter) {  // unable to find setter
+        continue;
+      }
+      switch (setter->getType()) {
```

Comment:
This is a duplicate of the logic in `hasDependencies` method. Pls extract this as a helper method.